### PR TITLE
Add Array.prototype.findLast() and Array.prototype.findLastIndex()

### LINKED ISSUE #48829: Add `Array.prototype.findLast` and `Array.prototype.findLastIndex`
# lib Update Request

## Configuration Check

My compilation *target* is `ES5` and my *lib* is `ESNext`.

## Missing / Incorrect Definition

Missing `Array.prototype.findLast` and `Array.prototype.findLastIndex`

## Sample Code

```ts
[3, 2, 1].findLast(x => x > 2);
```

## Documentation Link

Proposal: https://tc39.es/proposal-array-find-from-last/ (currently stage ~~3~~ 4)
Can I Use: https://caniuse.com/mdn-javascript_builtins_array_findlast (Chrome and Safari already shipped it without flags)

### PR DESCRIPTION
**Pull request by m-basov**
**Describe your changes**
Add type definitions for `Array.prototype.findLast()` and `Array.prototype.findLastIndex()` methods. Both methods were added to the `es2023` lib but the lib didn't exist so I've created it. The definitions are identical to `.find()` and `.findIndex()` but with updated comments.

**Testing performed**
Added `findLast.ts` test file where all Array interfaces try to call `.findLast()` and `.findLastIndex()`. 
Also, tested it manually by running:
- `tsc` with the `es5` target and `libs: []`
- `tsc` with the `es5` target and `libs: ["es2023"]`
- `tsc` with the `es5` target and `libs: ["esnext"]`

**Additional context**
The `es2023` target were added using this PR https://github.com/microsoft/TypeScript/pull/46291 as an example.

Fixes #48829

Credits to Bloomberg colleagues who helped with reviews @acutmore @dragomirtitian @mkubilayk.

### CODE REVIEW

**Comment by mkubilayk** [comment]
@DanielRosenwasser Has there been any changes to the CLA bot? I believe it used to accept contributions from the Bloomberg fork automatically.

File: src/compiler/types.ts
Comment (sandersn) [inline src/compiler/types.ts:7017]: this now overlaps with `ContainsTypeScriptClassSyntax = 1 << 13`, which means we're officially out of space in TransformFlags. =(

Paging @rbuckton for advice.

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good except that TransformFlags ran out of space. @rbuckton, do you have any ideas for how to move any other members out of it?

File: tests/cases/compiler/findLast.ts
Comment (rbuckton) [inline tests/cases/compiler/findLast.ts:3]: Can you add a test for the generic overload of `findLast`?

File: src/compiler/types.ts
Comment (rbuckton) [inline src/compiler/types.ts:6526]: @DanielRosenwasser: Do we normally add a new script target before that edition is ratified, or do we put everything in `esnext` until there's an official spec?

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by DanielRosenwasser** [comment]
I believe we wait for ratification because the target isn't stable.

**Comment by sandersn** [comment]
I'd vote to wait to add es2023 until that edition is ratified. It could go into esnext for now.

**Comment by m-basov** [comment]
I removed ES2023 target and left ES2023 lib as @DanielRosenwasser suggested. Could you take one more look @sandersn @rbuckton please? Thanks!

**Review by sandersn — APPROVED** [review]
Looks good. Can you merge from main so before I merge it?

**Comment by m-basov** [comment]
@sandersn done! Rebased on main.

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [inline src/lib/es2023.array.d.ts:11]: I don't think `this` should be typed `void`, that's probably an error in `es2015.core.d.ts`. I fixed it in my PR (#50450).
And I suggest repeating the doc too.

```suggestion
    findLast<S extends T>(predicate: (value: T, index: number, array: T[]) => value is S, thisArg?: any): S | undefined;
```

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [inline src/lib/es2023.array.d.ts:36]:
```suggestion
    findLast<S extends T>(predicate: (value: T, index: number, array: readonly T[]) => value is S, thisArg?: any): S | undefined;
```

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [inline src/lib/es2023.array.d.ts:61]: This overload didn't appear in `es5.d.ts` at all. Anyway it should be typed:
```suggestion
    findLast<S extends number>(predicate: (value: number, index: number, array: Int8Array) => value is S, thisArg?: any): S | undefined;
```
And so on.

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [inline src/lib/es2023.array.d.ts:62]:
```suggestion
    findLast(predicate: (value: number, index: number, array: Int8Array) => unknown, thisArg?: any): number | undefined;
```
and so on. Don't forget to repeat the doc.

**Comment by graphemecluster** [comment]
(I am not a member of TypeScript; I just reviewed based on my experience. Do point out me if I made something wrong.)

**Comment by robpalme** [comment]
Current status:  @m-basov is waiting on https://github.com/microsoft/TypeScript/pull/50450 by @graphemecluster to be merged.  Once that lands he will rebase this PR.

**Comment by graphemecluster** [comment]
@robpalme Thanks, I've just resolved the conflicts of my PRs.
And sorry if I interpret this incorrectly, but if Bloomberg would like to see my PRs being merged, please do anything that speeds up the review process, or otherwise anything that could get it started quicker, as it has been a while since I work on them. Appreciate your support on it!

**Comment by DanielRosenwasser** [comment]
Can we just implement this based on the existing signatures of `find` and `findIndex`, and then handle the changes to `findLast*` in #50450? That one is a much harder change to evaluate w.r.t. ecosystem impact, and there's no reason to block this on it.

**Comment by graphemecluster** [comment]
@DanielRosenwasser I suppose the only difference is whether or not the `this` in the predicate should be typed `void` (and whether to repeat the documentation). The others are, most likely, errors not existing in the current `find*` signature.

**Comment by m-basov** [comment]
Sure, I will rebase my changes on the latest main branch.

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [reply src/lib/es2023.array.d.ts:11]: FYI it’s now fixed by #51278.

Comment (m-basov) [reply src/lib/es2023.array.d.ts]: Thanks, I will fix the PR to match `.find()` declarations.

File: src/compiler/types.ts
Comment (DanielRosenwasser) [reply src/compiler/types.ts]: I think it's okay to put it into 2023 since we're expecting ratification before the final release of TS5.0.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Looks like we have some merge conflicts. Would you be able to fix those up?

Also, @rbuckton would you be able to take a second look?

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test top100
@typescript-bot run dt

File: src/lib/es2023.array.d.ts
Comment (graphemecluster) [inline src/lib/es2023.array.d.ts:112]:
```suggestion
    findLast(predicate: (value: number, index: number, array: Uint8ClampedArray) => unknown, thisArg?: any): number | undefined;

```

**Comment by m-basov** [comment]
@DanielRosenwasser thanks for review. I've fixed the merge conflicts.

@graphemecluster thanks for spotting the issue with `Uint8ClampedArray[]`. It's fixed now but I cannot find your comment in the GH UI (I noticed it on email).

**Comment by m-basov** [comment]
@microsoft-github-policy-service agree company="Bloomberg"

**Review by rbuckton — APPROVED** [review]
I'm a bit wary about adding an `es2023` lib given that the committee has not yet ratified ES2023, but given that Array find from last is at Stage 4, this is probably fine.

**Pull request merged**
